runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -1184,4 +1184,104 @@ function LanguageTree:language_for_range(range)
   return self
 end
 
+--- Get highlight information from parser.
+---
+---@return {group:string, range:integer[]}
+function LanguageTree:get_highlight()
+  ---@type {group:string,range:integer[],priority:integer,type:'start'|'end',start?:table,subpriority:integer}[]
+  local positions = {}
+  self:for_each_tree(function(tree, ltree)
+    local source = ltree:source()
+    local hl_query = query.get(ltree:lang(), 'highlights')
+
+    if not hl_query then
+      return
+    end
+
+    for id, node, metadata in hl_query:iter_captures(tree:root(), source, 0, -1) do
+      local name = hl_query.captures[id]
+      local start_row, start_col, end_row, end_col = node:range()
+      local priority = tonumber(metadata.priority or vim.highlight.priorities.treesitter)
+
+      table.insert(positions, {
```

Comment:
Is there a reason we need to inject two elements into `positions` instead of just providing the full range?

Reply:
That's how the algorithm works.

The algorithm resolves node priorities so that no node's range overlaps with another.

More about the algorithm:
A list of positions is prepared where each element is either a node's beginning or end position.
The positions are sorted by position (this separates the beginning and end positions of the same node).
Loop over the positions:
When encountering a beginning position its associated node gets appended to a priority sorted list (higher priority positions always come before lower priority positions).
When encountering an ending position its associated node gets removed from the priority sorted list.
On every cycle, if there is distance between the positions, the first node (e.g. the node with the highest priority) on the priority sorted list gets appended to the highlights list with its range set to the range between the previous position in the loop and the current position in the loop.

Reply:
These level of explanations work really well as code documentation.